awx/main/tasks/receptor.py

```diff
@@ -632,53 +632,21 @@ def kube_config(self):
         return config
 
 
-# TODO: receptor reload expects ordering within config items to be preserved
-# if python dictionary is not preserving order properly, may need to find a
-# solution. yaml.dump does not seem to work well with OrderedDict. below line may help
-# yaml.add_representer(OrderedDict, lambda dumper, data: dumper.represent_mapping('tag:yaml.org,2002:map', data.items()))
-#
-RECEPTOR_CONFIG_STARTER = (
-    {'local-only': None},
-    {'log-level': 'info'},
-    {'node': {'firewallrules': [{'action': 'reject', 'tonode': settings.CLUSTER_HOST_ID, 'toservice': 'control'}]}},
-    {'control-service': {'service': 'control', 'filename': '/var/run/receptor/receptor.sock', 'permissions': '0660'}},
-    {'work-command': {'worktype': 'local', 'command': 'ansible-runner', 'params': 'worker', 'allowruntimeparams': True}},
-    {'work-signing': {'privatekey': '/etc/receptor/signing/work-private-key.pem', 'tokenexpiration': '1m'}},
-    {
-        'work-kubernetes': {
-            'worktype': 'kubernetes-runtime-auth',
-            'authmethod': 'runtime',
-            'allowruntimeauth': True,
-            'allowruntimepod': True,
-            'allowruntimeparams': True,
-        }
-    },
-    {
-        'work-kubernetes': {
-            'worktype': 'kubernetes-incluster-auth',
-            'authmethod': 'incluster',
-            'allowruntimeauth': True,
-            'allowruntimepod': True,
-            'allowruntimeparams': True,
-        }
-    },
-    {
-        'tls-client': {
-            'name': 'tlsclient',
-            'rootcas': '/etc/receptor/tls/ca/mesh-CA.crt',
-            'cert': '/etc/receptor/tls/receptor.crt',
-            'key': '/etc/receptor/tls/receptor.key',
-            'mintls13': False,
-        }
-    },
-)
-
-
 @task()
 def write_receptor_config():
+    current_config = read_receptor_config()  # this grabs lock
```

Comment:
Isn't this dangerous? By using the current config as a starter, aren't you kind of cargo-culting the config in an automated way? That could create some weird problems of config drift down the line.

Also, I'd like to move in the direction of generating `receptor_config` in a separate method so that we can test it. Before this, it looked like all we needed to test it was the `Instance` and `InstanceLink` objects in the database. We can still test this after your restructuring here, but that just means moving the "starter config" to the tests.

Reply:
Looking down at the line `receptor_config.append(peer)`, I can see this not working _at all_ since you're not checking in advance whether the link already exists in the config, so it could end up adding it every time it rewrites the config.

Reply:
**cargo culting** (uncountable) (chiefly computing) An approach that copies an existing successful approach without properly analysing and understanding it.

Reply:
The logic here strips out any "peer" entry, and we simply readd them later. So in a way we are starting "fresh" without having the hard coded template.

I think I will just revert this part back for now and we can figure out a better solution later.